File: qt/multimedia/gen_qaudiooutput.cpp
// +build ignore

#include <QAudioDevice>
#include <QAudioOutput>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qaudiooutput.h>
#include "gen_qaudiooutput.h"
class MiqtVirtualQAudioOutput : public virtual QAudioOutput {
public:
MiqtVirtualQAudioOutput(): QAudioOutput() {};
MiqtVirtualQAudioOutput(const QAudioDevice& device): QAudioOutput(device) {};
MiqtVirtualQAudioOutput(QObject* parent): QAudioOutput(parent) {};
MiqtVirtualQAudioOutput(const QAudioDevice& device, QObject* parent): QAudioOutput(device, parent) {};

virtual ~MiqtVirtualQAudioOutput() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QAudioOutput::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QAudioOutput_MetaObject(const_cast<MiqtVirtualQAudioOutput*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QAudioOutput::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QAudioOutput::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QAudioOutput_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QAudioOutput::qt_metacast(param1);

	}
};
QAudioOutput* QAudioOutput_new() {
return new MiqtVirtualQAudioOutput();
}
QAudioOutput* QAudioOutput_new2(QAudioDevice* device) {
return new MiqtVirtualQAudioOutput(*device);
}
QAudioOutput* QAudioOutput_new3(QObject* parent) {
return new MiqtVirtualQAudioOutput(parent);
}
QAudioOutput* QAudioOutput_new4(QAudioDevice* device, QObject* parent) {
return new MiqtVirtualQAudioOutput(*device, parent);
}
void QAudioOutput_virtbase(QAudioOutput* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QAudioOutput_MetaObject(const QAudioOutput* self) {
	return (QMetaObject*) self->metaObject();
}
void* QAudioOutput_Metacast(QAudioOutput* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QAudioOutput_Tr(const char* s) {
	QString _ret = QAudioOutput::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
QAudioDevice* QAudioOutput_Device(const QAudioOutput* self) {
	return new QAudioDevice(self->device());
}
float QAudioOutput_Volume(const QAudioOutput* self) {
	return self->volume();
}
bool QAudioOutput_IsMuted(const QAudioOutput* self) {
	return self->isMuted();
}
void QAudioOutput_SetDevice(QAudioOutput* self, QAudioDevice* device) {
	self->setDevice(*device);
}
void QAudioOutput_SetVolume(QAudioOutput* self, float volume) {
	self->setVolume(static_cast<float>(volume));
}
void QAudioOutput_SetMuted(QAudioOutput* self, bool muted) {
	self->setMuted(muted);
}
void QAudioOutput_DeviceChanged(QAudioOutput* self) {
	self->deviceChanged();
}
void QAudioOutput_connect_DeviceChanged(QAudioOutput* self, intptr_t slot) {
	MiqtVirtualQAudioOutput::connect(self, static_cast<void (QAudioOutput::*)()>(&QAudioOutput::deviceChanged), self, [=]() {
		miqt_exec_callback_QAudioOutput_DeviceChanged(slot);
	});
}
void QAudioOutput_VolumeChanged(QAudioOutput* self, float volume) {
	self->volumeChanged(static_cast<float>(volume));
}
void QAudioOutput_connect_VolumeChanged(QAudioOutput* self, intptr_t slot) {
	MiqtVirtualQAudioOutput::connect(self, static_cast<void (QAudioOutput::*)(float)>(&QAudioOutput::volumeChanged), self, [=](float volume) {
		float sigval1 = volume;
		miqt_exec_callback_QAudioOutput_VolumeChanged(slot, sigval1);
	});
}
void QAudioOutput_MutedChanged(QAudioOutput* self, bool muted) {
	self->mutedChanged(muted);
}
void QAudioOutput_connect_MutedChanged(QAudioOutput* self, intptr_t slot) {
	MiqtVirtualQAudioOutput::connect(self, static_cast<void (QAudioOutput::*)(bool)>(&QAudioOutput::mutedChanged), self, [=](bool muted) {
		bool sigval1 = muted;
		miqt_exec_callback_QAudioOutput_MutedChanged(slot, sigval1);
	});
}
struct miqt_string QAudioOutput_Tr2(const char* s, const char* c) {
	QString _ret = QAudioOutput::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QAudioOutput_Tr3(const char* s, const char* c, int n) {
	QString _ret = QAudioOutput::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QAudioOutput_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioOutput*>( (QAudioOutput*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QAudioOutput_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQAudioOutput*)(self) )->virtualbase_MetaObject();
}
void QAudioOutput_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioOutput*>( (QAudioOutput*)(self) )->handle__Metacast = slot;
}
void* QAudioOutput_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQAudioOutput*)(self) )->virtualbase_Metacast(param1);
}
void QAudioOutput_Delete(QAudioOutput* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQAudioOutput*>( self );
	} else {
		delete self;
	}
}

File: qt/multimedia/gen_qaudiooutput.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QAudioOutput struct {
	h          uintptr
	isSubclass bool
}

// NewQAudioOutput constructs a new QAudioOutput object.
func NewQAudioOutput() *QAudioOutput {
	g := newQAudioOutput(QAudioOutput_new())
	g.isSubclass = true
	return g
}

// NewQAudioOutput2 constructs a new QAudioOutput object.
func NewQAudioOutput2(device *QAudioDevice) *QAudioOutput {
	g := newQAudioOutput(QAudioOutput_new2(device.cPointer()))
	g.isSubclass = true
	return g
}

// NewQAudioOutput3 constructs a new QAudioOutput object.
func NewQAudioOutput3(parent *qt.QObject) *QAudioOutput {
	g := newQAudioOutput(QAudioOutput_new3((*QObject)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

// NewQAudioOutput4 constructs a new QAudioOutput object.
func NewQAudioOutput4(device *QAudioDevice, parent *qt.QObject) *QAudioOutput {
	g := newQAudioOutput(QAudioOutput_new4(device.cPointer(), (*QObject)(parent.UnsafePointer())))
	g.isSubclass = true
	return g
}

func (this *QAudioOutput) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QAudioOutput_MetaObject(this.h)))
}

func (this *QAudioOutput) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QAudioOutput_Metacast(this.h, param1_Cstring))
}

func QAudioOutput_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QAudioOutput_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioOutput) Device() *QAudioDevice {
	_goptr := newQAudioDevice(QAudioOutput_Device(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAudioOutput) Volume() float32 {
	return (float32)(QAudioOutput_Volume(this.h))
}

func (this *QAudioOutput) IsMuted() bool {
	return (bool)(QAudioOutput_IsMuted(this.h))
}

func (this *QAudioOutput) SetDevice(device *QAudioDevice) {
	QAudioOutput_SetDevice(this.h, device.cPointer())
}

func (this *QAudioOutput) SetVolume(volume float32) {
	QAudioOutput_SetVolume(this.h, (float)(volume))
}

func (this *QAudioOutput) SetMuted(muted bool) {
	QAudioOutput_SetMuted(this.h, (bool)(muted))
}

func (this *QAudioOutput) DeviceChanged() {
	QAudioOutput_DeviceChanged(this.h)
}

func (this *QAudioOutput) OnDeviceChanged(slot func()) {
	QAudioOutput_connect_DeviceChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioOutput_DeviceChanged
func miqt_exec_callback_QAudioOutput_DeviceChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QAudioOutput) VolumeChanged(volume float32) {
	QAudioOutput_VolumeChanged(this.h, (float)(volume))
}

func (this *QAudioOutput) OnVolumeChanged(slot func(volume float32)) {
	QAudioOutput_connect_VolumeChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioOutput_VolumeChanged
func miqt_exec_callback_QAudioOutput_VolumeChanged(cb intptr_t, volume float) {
	gofunc, ok := cgo.Handle(cb).Value().(func(volume float32))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (float32)(volume)

	gofunc(slotval1)
}

func (this *QAudioOutput) MutedChanged(muted bool) {
	QAudioOutput_MutedChanged(this.h, (bool)(muted))
}

func (this *QAudioOutput) OnMutedChanged(slot func(muted bool)) {
	QAudioOutput_connect_MutedChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioOutput_MutedChanged
func miqt_exec_callback_QAudioOutput_MutedChanged(cb intptr_t, muted bool) {
	gofunc, ok := cgo.Handle(cb).Value().(func(muted bool))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (bool)(muted)

	gofunc(slotval1)
}

func QAudioOutput_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioOutput_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QAudioOutput_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioOutput_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioOutput) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QAudioOutput_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QAudioOutput) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioOutput_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioOutput_MetaObject
func miqt_exec_callback_QAudioOutput_MetaObject(self QAudioOutput, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QAudioOutput{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QAudioOutput) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QAudioOutput_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QAudioOutput) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioOutput_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioOutput_Metacast
func miqt_exec_callback_QAudioOutput_Metacast(self QAudioOutput, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QAudioOutput{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
